Adding conformance checks

VestigeJ opened the pull request:
<!-- HTML Comments can be left in place or removed. -->

#### Proposed Changes ####

<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. -->

add support for running conformance checks automatically after all of the terraform busywork

<!-- Does this change require an update to documentation? -->

yes added to the docs - basically make sure you supply at least 1 agent and 1 server for this test to run successfully.
#### Types of Changes ####

<!-- What types of changes does your code introduce to distro framework? Issue validation, Patch Validation, Fix, New functionality, Refactor or etc -->

#### Testing ####
<!-- Answer the checklist bellow  -->

Checklist:
1. If your PR changes anything on or related to Jenkins, run it pointing to your branch to make sure it's okay.


2. Verify code lint; we should not have errors.
This PR also turns shellcheck on for more folders specifically the scripts/ directory where it should've been on for awhile now. I expect shellcheck to fail a bit but I'll take the work up for fixing those after this PR gets merged.

3. Update the documentation if needed.
🟢 

4. Update makefile and docker run if adding new tests.
🟢 

5. Run your tests at least 4 times with all configurations needed and possible.
🟢 

6. If needed test with different os types.
🟢 

#### Linked Issues ####

<!-- Link any related issues, pull-requests, qa-tasks repo issues or commit hashes that are relevant to this pull request. If you are opening a PR without a corresponding issue please consider creating one first, at https://github.com/rancher/distros-test-framework/issues . -->

#### Further Comments ####

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

fmoral2 commented on pkg/testcase/conformance.go:
```diff
@@ -0,0 +1,138 @@
+	}
+}
+
+func ConformanceTest(testName string) {
```
do u think its better to read like: 
```

func ConformanceTest(testName string) {
	installConformanceBinary()
	launchSonobuoyTests(testName)

	checkStatus()
	testResultTar := getResults()
	shared.LogLevel("info", "%s", "testResultTar: "+testResultTar)

	rerunFailedTests(testResultTar)

	testResultTar = getResults()
	shared.LogLevel("info", "%s", "testResultTar: "+testResultTar)

	parseResults(testResultTar)

	cleanupTests()
}

```
 this way who is reading have a more clear picture of whats happening because its easier to read and gets block separated on what each step is doing ?

fmoral2 replied:
same for other functions

VestigeJ replied:
Originally I didn't want to alter MD's mixed OS test as it was meant to be a small PR to add conformance checks not alter the existing test.

fmoral2 commented on scripts/install_sonobuoy.sh:
```diff
@@ -0,0 +1,72 @@
+}
+
+deletion(){
+    #has_bin sonobuoy
```
if u are not using, can delete this func `has_bin`

VestigeJ replied:
I left this commented in so that when I take the shellchecks to fix the scripts directory I can add it in to check for the existence of a binary.

fmoral2 commented on entrypoint/conformance/sonobuoy_test.go:
```diff
@@ -0,0 +1,44 @@
+	})
+
+	It("Validates the releases conformance with upstream requirements", func() {
+		testcase.ConformanceTest("certified-conformance")
```
if you are not using the parameter here to actually pass down a parameter option, its useless right? 

just call your test type "certified-conformance" directly on its usage on the `launchSonobuoyTests`

VestigeJ replied:
Similar to the commented out has_bin function for scripts this structure is for future functionality to be added with sonobuoy running different existing tests available from upstream. Specifically sig-networking as an example or a "quick" conformance check.

VestigeJ replied:
The downside of relying heavily on tfvars cfg is that it makes this test suite more tightly coupled to Terraform.

VestigeJ commented:
Hey @fmoral2 FYI this prod here wasn't used because when you pass the image version to sonobuoy it's expecting the kubernetes image version NOT the k3s/rke2 version. 

The sonobuoy CLI expects the upstream kubernetes version 
 --kubernetes-version="${_version}"

```
unc ExtractKubeImageVersion() string {
 	_, serverVersion, err := Product()
 	prod, serverVersion, err := Product()
 	if err != nil {
 		LogLevel("error", "error retrieving version of product: %s", err)
 		os.Exit(1)
 	}
 
 	version := strings.Split(serverVersion, "+")[0]
 	version = strings.TrimPrefix(version, prod+" version ")
 	version = strings.TrimSpace(version)
 
 	if strings.Contains(version, "-rc") {
 		version = strings.Split(version, "-rc")[0]
 	}
 
 	if version == "" {
 		LogLevel("error", "%s failed to resolve to server version string: %s", serverVersion, err)
 		os.Exit(1)
```

fmoral2 commented:
> Hey @fmoral2 FYI this prod here wasn't used because when you pass the image version to sonobuoy it's expecting the kubernetes image version NOT the k3s/rke2 version.
> 
> The sonobuoy CLI expects the upstream kubernetes version --kubernetes-version="${_version}"
> 
> ```
> unc ExtractKubeImageVersion() string {
>  	_, serverVersion, err := Product()
>  	prod, serverVersion, err := Product()
>  	if err != nil {
>  		LogLevel("error", "error retrieving version of product: %s", err)
>  		os.Exit(1)
>  	}
>  
>  	version := strings.Split(serverVersion, "+")[0]
>  	version = strings.TrimPrefix(version, prod+" version ")
>  	version = strings.TrimSpace(version)
>  
>  	if strings.Contains(version, "-rc") {
>  		version = strings.Split(version, "-rc")[0]
>  	}
>  
>  	if version == "" {
>  		LogLevel("error", "%s failed to resolve to server version string: %s", serverVersion, err)
>  		os.Exit(1)
> ```



sure, but thats exactly why am using now 
because this fucntion  returns like
```

_, serverVersion, err := Product()

serverVersion = === > k3s version v1.32.3+k3s1

```

so am using the product name to trim it out as showed on code

so we will ended up with `v1.32.3`

mdrahman-suse commented on Makefile:
```diff
@@ -114,14 +117,16 @@ test-system-default-registry:
 
 #========================= TestCode Static Quality Check =========================#
-pre-commit: go-check shell-check
+pre-commit: go-check
```
Why are we removing shell-check from here?

VestigeJ replied:
Adding in a later PR where I will also overhaul the scripts directory contents.

mdrahman-suse commented on Makefile:
```diff
@@ -114,14 +117,16 @@ test-system-default-registry:
 shell-check:
 	@shellcheck modules/airgap/setup/*.sh
 	@shellcheck modules/ipv6only/scripts/*.sh
+	@shellcheck scripts/*.sh
```
nit: line of EOF

mdrahman-suse commented on docs/testing.md:
```diff
@@ -284,4 +284,21 @@ S3_FOLDER=snapshots
+- Selinux Setup
+
+## Validating Conformance Tests with Sonobuoy
+ - Please note that the sonobuoy version has not been updated for a year and the functionality of sonobuoy is degrading with minor versions of k8s.
```
Can we add this on a NOTE section?

mdrahman-suse commented on docs/testing.md:
```diff
@@ -284,4 +284,21 @@ S3_FOLDER=snapshots
+ - Please note that the sonobuoy version has not been updated for a year and the functionality of sonobuoy is degrading with minor versions of k8s.
+ - Full conformance tests done for patch validations should be 3 servers 1 agent minimum.
+ - You can use the make file command `make test-run-sonobuoy` to run the conformance tests.
+ - Additionally you can use `go test -timeout=140m -v -count=1 ./entrypoint/conformance/... --ginkgo.timeout=140m` you must extend the ginkgo timeout in addition to the go test timeout 
```
is this the minimum timeout? does timeout vary based on the product?

VestigeJ replied:
This timeout was for RKE2, I did notice Fran expanded it even larger perhaps to accommodate running through Jenkins.

mdrahman-suse commented on entrypoint/conformance/conformance_test.go:
```diff
@@ -0,0 +1,43 @@
+			assert.PodAssertReady())
+	})
+
+	It("Validates the releases conformance with upstream requirements", func() {
```
nit: `the release is conformant with`

mdrahman-suse commented on pkg/testcase/conformance.go:
```diff
@@ -0,0 +1,301 @@
+
+	_ = getResults(newTar)
+
+	cleanupTests()
```
Can we add some debug logging here?

mdrahman-suse commented on scripts/install_sonobuoy.sh:
```diff
@@ -0,0 +1,71 @@
+action=${1:-install}
+version=${2:-0.57.3}
+arch=${3:-amd64}
+mixed_plugins_url="git clone https://github.com/phillipsj/my-sonobuoy-plugins.git"
```
Should think about moving away from this repo

mdrahman-suse reviewed (approved):
Few comments, non-blocking can be addressed in later PR

mdrahman-suse commented on entrypoint/mixedoscluster/mixedoscluster_suite_test.go:
```diff
@@ -19,13 +19,15 @@ var (
-	flag.StringVar(&customflag.ServiceFlag.External.SonobuoyVersion, "sonobuoyVersion", "0.56.17", "Sonobuoy Version that will be executed on the cluster")
-	flag.Var(&customflag.ServiceFlag.Destroy, "destroy", "Destroy cluster after test")
+	flags = &customflag.ServiceFlag
+	flag.StringVar(&flags.External.SonobuoyVersion, "sonobuoyVersion", "0.56.17", "Sonobuoy Version that will be executed on the cluster")
```
should update the default to the latest
